vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher.go

```diff
@@ -880,6 +882,15 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
 	case curObjPasses && oldObjPasses:
 		watchEvent = watch.Event{Type: watch.Modified, Object: object}
 	case !curObjPasses && oldObjPasses:
+		// return a delete event with the previous object content, but with the event's resource version
+		object, err := c.copier.Copy(event.PrevObject)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("unexpected copy error: %v", err))
+			return
+		}
+		if err := c.versioner.UpdateObject(object, event.ResourceVersion); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failure to version api object (%d) %#v: %v", event.ResourceVersion, object, err))
+		}
 		watchEvent = watch.Event{Type: watch.Deleted, Object: object}
```

Comment:
this looks like it was completely wrong before, not sending PrevObject in delete events...

Reply:
https://github.com/kubernetes/kubernetes/commit/e9e69356e4907fa4d0f45ea7e7768357ba71aba9

Reply:
move `object, err := c.copier.Copy(event.Object)` into the added/modified cases then?

Reply:
and update the description to indicate this is a pick of those two fixes